src/xas_tdp_correction.F

```diff
@@ -0,0 +1,1870 @@
+   USE kinds,                           ONLY: dp
+   USE machine,                         ONLY: m_flush
+   USE message_passing,                 ONLY: mp_sum
+   USE physcon,                         ONLY: evolt
+   USE qs_environment_types,            ONLY: get_qs_env,&
+                                              qs_environment_type
+   USE qs_ks_methods,                   ONLY: qs_ks_build_kohn_sham_matrix
+   USE qs_mo_types,                     ONLY: deallocate_mo_set,&
+                                              duplicate_mo_set,&
+                                              get_mo_set,&
+                                              mo_set_p_type,&
+                                              reassign_allocated_mos
+   USE util,                            ONLY: get_limit
+   USE xas_tdp_kernel,                  ONLY: contract2_AO_to_doMO,&
+                                              ri_all_blocks_mm
+   USE xas_tdp_types,                   ONLY: donor_state_type,&
+                                              xas_tdp_control_type,&
+                                              xas_tdp_env_type
+
+!$ USE OMP_LIB, ONLY: omp_get_max_threads, omp_get_thread_num
+#include "./base/base_uses.f90"
+
+   IMPLICIT NONE
+   PRIVATE
+
+   CHARACTER(len=*), PARAMETER, PRIVATE :: moduleN = 'xas_tdp_correction'
+
+   PUBLIC :: gw2x_shift, get_soc_splitting
+
+CONTAINS
+
+! **************************************************************************************************
+!> \brief First implementation of the very legit GW2X correction of the Kohn-Sham IP from
+!>        https://onlinelibrary.wiley.com/doi/pdf/10.1002/qua.1543
+!> \param donor_state ...
+!> \param xas_tdp_env ...
+!> \param xas_tdp_control ...
+!> \param qs_env ...
+! **************************************************************************************************
+   SUBROUTINE GW2X_shift(donor_state, xas_tdp_env, xas_tdp_control, qs_env)
+
+      TYPE(donor_state_type), POINTER                    :: donor_state
+      TYPE(xas_tdp_env_type), POINTER                    :: xas_tdp_env
+      TYPE(xas_tdp_control_type), POINTER                :: xas_tdp_control
+      TYPE(qs_environment_type), POINTER                 :: qs_env
+
+      CHARACTER(len=*), PARAMETER :: routineN = 'GW2X_shift', routineP = moduleN//':'//routineN
+
+      INTEGER :: ex_idx, exat, first_domo(2), handle, i, ido_mo, iloc, ilocat, ispin, jspin, &
+         locat, nao, natom, ndo_mo, nhomo(2), nlumo(2), nonloc, nspins, start_sgf
+      INTEGER, DIMENSION(:), POINTER                     :: nsgf_blk
+      LOGICAL                                            :: pseudo_canonical
+      REAL(dp)                                           :: og_hfx_frac
+      REAL(dp), ALLOCATABLE, DIMENSION(:, :)             :: contract_coeffs_backup
+      TYPE(admm_type), POINTER                           :: admm_env
+      TYPE(cp_1d_r_p_type), DIMENSION(:), POINTER        :: homo_evals, lumo_evals
```

Comment:
`ALLOCATABLE` instead and change the interface for `GW2X_os_iterations` to take it just as a normal array?

Reply:
This would be a pain in open-shell calculations, I think. Carrying the eigenvalues this way allows to iterate over the spin and access the relevant eigenvalues when needed. The alternative would be to explicitelly have and alpha and beta version of both arrays, which I do not like.

Reply:
How so? You just need a dynamically allocated array of type `cp_1d_r_p_type`, correct? And you are using it only in that procedure without having to call another API which would require it to be a POINTER or TARGET.

Reply:
I might have missunderstood. You mean `TYPE(cp_1d_r_p_type), DIMENSION(:), ALLOCATABLE  :: homo_evals, lumo_evals` ? In that case, yes, there is nothing preventing me from doing that and I will change it.

Reply:
yes, that :)